Update GPX export path to Documents folder for user accessibility

Andyporras opened the pull request:
This pull request modifies the GPX export path to ensure files are saved in a visible and accessible location for the user. Previously, files were exported to a default location that was not accessible, making it difficult for users to locate them. With this update, GPX files are saved in the Downloads folder, improving the user experience.

Compatibility: The solution has been tested and works up to the latest version of Android 14.

Important Note: This PR depends on the approval of #457 'Updating the target SDK to 35 with the necessary changes in dependencies, permissions and execution flows'. Please make sure to approve PR #457  before proceeding with this one."

Issues solved:
>*  #443  
>* #440

Andyporras commented:
> Esta solicitud de extracción modifica la ruta de exportación de GPX para garantizar que los archivos se guarden en una ubicación visible y accesible para el usuario. Anteriormente, los archivos se exportaban a una ubicación predeterminada que no era accesible, lo que dificultaba la localización de los usuarios. Con esta actualización, los archivos GPX se guardan en la carpeta Descargas, mejorando la experiencia del usuario.
> 
> Compatibilidad: La solución ha sido probada y funciona hasta la última versión de Android 14.
> 
> Nota importante: Esta PR depende de la aprobación de #457 'Actualización del SDK de destino a 35 con los cambios necesarios en dependencias, permisos y flujos de ejecución'. Asegúrese de aprobar PR #457 antes de proceder con este."
> 
> Problemas resueltos:
> 
> > * [La exportación GPX no funciona #443](https://github.com/labexp/osmtracker-android/issues/443)
> > * [OSM Tracker GPS Exportación rota #440](https://github.com/labexp/osmtracker-android/issues/440)

I wanted to provide some context regarding this PR, "Update GPX export path to Downloads folder for user accessibility #458," in relation to PR #443 ("GPX export not working").

While both PRs address the GPX export functionality, there are a few key differences:

Export Location: This PR updates the export path to the Downloads folder for improved user accessibility, whereas PR #443 saves files in the Documents directory.

UI Message Update: This PR also modifies the message to accurately reflect the new export location, ensuring better clarity for users.

That said, if PR #443 is accepted as is, this PR may not be necessary, as both resolve the core issue of the GPX export functionality. However, I believe the choice of export location (Downloads vs. Documents) and the updated UI message could be factors to consider in making the final decision.

Let me know if further clarification or adjustments are needed!

FR3DD221 commented:
Works perfectly! @Andyporras

Kevin-Salazar-itcr commented:
Hello!
It works successfully! TY @Andyporras

Binnette commented:
Hi @Andyporras Can you please update your PR. You have some conflicts with the previous PR #451 which was already addressing (part?) of this issue.

Binnette commented:
In PR #451 storage have been set to DIRECTORY_DOCUMENTS. I think it is a way better idea than download. As download can be messy due to user many downloading files from web, etc.

Andyporras commented:
> En PR #451 el almacenamiento se ha establecido en DIRECTORY_DOCUMENTS. Creo que es una idea mucho mejor que descargar. Como la descarga puede ser desordenada debido a que el usuario descarga muchos archivos de la web, etc.

Thank you for pointing this out. I’ll review the conflicts with PR #451 and make the necessary updates to ensure compatibility.

If we agree that DIRECTORY_DOCUMENTS is a better location, I can adjust this PR to align with the changes introduced in PR #451. I’ll update the implementation accordingly and re-test to ensure everything works seamlessly.

I’ll notify you once the conflicts are resolved and the PR is updated. Thank you for the feedback!

Andyporras commented:
@Binnette 

I’ve resolved the conflicts and updated the logic to ensure that GPX files are saved in the Documents directory. Additionally, I’ve updated the UI message to "Public directory (in documents)" to make it clear to the user where the files are being stored.

Let me know if you need any further adjustments!

Binnette commented on app/src/main/java/net/osmtracker/activity/About.java:
```diff
@@ -140,6 +140,8 @@ private String getDebugInfo() {
 				+ baseExportDirectory + "'\n";
+
+
 	}
```
2 blanks lines to delete

Binnette commented on app/src/main/java/net/osmtracker/gpx/ExportToStorageTask.java:
```diff
@@ -38,9 +38,11 @@ public ExportToStorageTask(Context context, long... trackId) {
 	protected File getExportDirectory(Date startDate) throws ExportTrackException {
         SharedPreferences preferences = PreferenceManager.getDefaultSharedPreferences(context);
 
+
```
blank line to delete

Binnette commented on app/src/main/java/net/osmtracker/gpx/ExportToStorageTask.java:
```diff
@@ -86,32 +88,33 @@ private boolean isExternalStorageWritable() {
+	public File getBaseExportDirectory(SharedPreferences prefs) throws ExportTrackException {
+		// Verify if folder exists and is writable
+		if (!Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState())) {
 			throw new ExportTrackException(
```
Nope. You should read code of function isExternalStorageWritable. This function already do the job.

Andyporras replied:
You're right! I reviewed the isExternalStorageWritable function, and it already handles the logic for ensuring that the storage is writable. Thanks for pointing that out, I’ll make sure the implementation reflects that.

Binnette commented on app/src/main/java/net/osmtracker/gpx/ExportToStorageTask.java:
```diff
@@ -86,32 +88,33 @@ private boolean isExternalStorageWritable() {
-    }
+	}
+
+
```
2 blank lines to delete

Binnette commented on app/src/main/java/net/osmtracker/layout/GpsStatusRecord.java:
```diff
@@ -169,6 +169,7 @@ else if (gpsActive && !activity.getButtonsEnabled()) {
 				manageRecordingIndicator(true);
 			}
 
+
```
1 blank line to delete

Binnette reviewed (changes requested):
Hi @Andyporras, please read my review. Also use tab "Files changes" on your PR to saw a sumup of your changes.

Binnette commented:
In the end, you implemented the same code as in PR #451. I will accept your PR for the logs, etc.

Binnette commented on app/src/main/res/values/strings-preferences.xml:
```diff
@@ -73,7 +73,7 @@
 
 	<string name="prefs_output">GPX settings</string>
-	<string name="prefs_storage_dir">External storage (SD) directory</string>
+	<string name="prefs_storage_dir">Public directory (in documents)</string>
```
Why not: "Storage folder in documents" ?

Because this setting allow user to set the name of the subfolder inside document folder where the GPX will be stored.

Andyporras replied:
That sounds good! I agree with the suggestion of using "Storage folder in documents." It accurately reflects that the setting allows the user to define the subfolder within the Documents directory where the GPX files will be stored.

Andyporras commented:
@Binnette I have already made the changes you suggested
